src/frontend/src/scheduler/snapshot.rs

```diff
@@ -93,17 +109,35 @@ impl std::fmt::Debug for PinnedSnapshot {
 pub type PinnedSnapshotRef = Arc<PinnedSnapshot>;
 
 impl PinnedSnapshot {
-    fn batch_query_epoch(&self, is_barrier_read: bool) -> BatchQueryEpoch {
-        let epoch = if is_barrier_read {
-            batch_query_epoch::Epoch::Current(u64::MAX)
-        } else {
-            batch_query_epoch::Epoch::Committed(self.value.max_committed_epoch)
-        };
-        BatchQueryEpoch { epoch: Some(epoch) }
+    fn batch_query_epoch(&self, scan_tables: &HashSet<TableId>) -> Result<Epoch, SchedulerError> {
+        // use the min committed epoch of tables involved in the scan
+        let epoch = scan_tables
+            .iter()
+            .map(|table_id| {
+                self.value
+                    .state_table_info
+                    .info()
+                    .get(table_id)
+                    .map(|info| Epoch(info.committed_epoch))
+                    .ok_or_else(|| anyhow!("table id {table_id} may have been dropped"))
+            })
+            .try_fold(None, |prev_min_committed_epoch, committed_epoch| {
+                committed_epoch.map(|committed_epoch| {
+                    if let Some(prev_min_committed_epoch) = prev_min_committed_epoch
+                        && prev_min_committed_epoch <= committed_epoch
+                    {
+                        Some(prev_min_committed_epoch)
+                    } else {
+                        Some(committed_epoch)
+                    }
+                })
+            })?
+            .unwrap_or_else(Epoch::now);
```

Comment:
Shall we use the maximum committed epoch of any table instead? This ensures that the timestamp is still the one that is used by some epoch, instead of locally generated one.